proposals/csharp-12.0/collection-expressions.md

```diff
@@ -874,6 +874,114 @@ However, given the breadth and consistency brought by the new literal syntax, we
+    public static void M1(char[] x)
+    {
+    }
+    
+    void Test()
+    {
+        M1(['a', 'b']); // error CS0121: The call is ambiguous between the following methods or properties: 'C1.M1(string)' and 'C1.M1(char[])'
+    }
+}
+```
+
+However, the 'C1.M1(string)' is not a candidate that can be used because:
+```
+error CS1729: 'string' does not contain a constructor that takes 0 arguments
+error CS1061: 'string' does not contain a definition for 'Add' and no accessible extension method 'Add' accepting a first argument of type 'string' could be found (are you missing a using directive or an assembly reference?)
+```
+
+Here is another example with a user-defined type and a stronger error that doesn't even mention a valid candidate:
+``` C#
+using System.Collections;
+using System.Collections.Generic;
+
+class C1 : IEnumerable<char>
+{
+    public static void M1(C1 x)
+    {
+    }
+    public static void M1(char[] x)
+    {
+    }
+
+    void Test()
+    {
+        M1(['a', 'b']); // error CS1061: 'C1' does not contain a definition for 'Add' and no accessible extension method 'Add' accepting a first argument of type 'C1' could be found (are you missing a using directive or an assembly reference?)
+    }
+
+    public static implicit operator char[](C1 x) => throw null;
+    IEnumerator<char> IEnumerable<char>.GetEnumerator() => throw null;
+    IEnumerator IEnumerable.GetEnumerator() => throw null;
+}
+```
+
+It looks like the situation is very similar to what we used to have with method group to delegate conversions. I.e.
+there were scenarios where the conversion existed, but was erroneous. We decided to improve that by ensuring that,
+if conversion is erroneous, then it doesn't exist.
+
+Note, that with "Params Collections" feature we will be running into a similar issue. It might be good to disallow
+usage of `params` modifier for not constructible collections. However in the current proposal that check is based on 
+[*conversions*](#conversions) section. Here is an example:
+
+``` C#
+using System.Collections;
+using System.Collections.Generic;
+
+class C1 : IEnumerable<char>
+{
+    public static void M1(params C1 x) // It is probably better to report an error about an invalid `params` modifier
+    {
+    }
+    public static void M1(params ushort[] x)
+    {
+    }
+
+    void Test()
+    {
+        M1('a', 'b'); // error CS1061: 'C1' does not contain a definition for 'Add' and no accessible extension method 'Add' accepting a first argument of type 'C1' could be found (are you missing a using directive or an assembly reference?)
+        M2('a', 'b'); // Ok
+    }
+
+    public static void M2(params ushort[] x)
+    {
+    }
+
+    IEnumerator<char> IEnumerable<char>.GetEnumerator() => throw null;
+    IEnumerator IEnumerable.GetEnumerator() => throw null;
+}
+```
+
+It looks like the issue was somewhat discussed previously, see https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-10-02.md#collection-expressions.
```

Comment:
I feel like, for the example you included at the start here, it's important to include another part of that same meeting:

> The second API, though, is not ambiguous by today's rules, as string is not constructable and therefore no conversion exists. By the new rules though, string is a valid target type for collection expressions, and will simply fail at a later stage. This makes the M2 API ambiguous. After some more consideration, we think that there's good reasoning for making strings constructable with collection expressions. List patterns already work on them, and we think that syntax could be useful for combining multiple string slices together. Support for this, though, is likely not going to make C# 12, and will be considered for future C# versions.

LDM intentionally considered that `string` and `char[]` overloads will be ambiguous with the current rules, and intentionally proceeded with the intention that we will at some point allow creation of strings with collection expressions.

Reply:
> I feel like, for the example you included at the start here, it's important to include another part of that same meeting

I do not think it makes sense to copy the meeting notes here. Especially that the issue is not string specific and I do not want to make it appear this way. If during discussion we decide that strings are special, that would be fine. However, that is not my proposal, which I am still working on formulating.